ethereum/client_test.go

```diff
@@ -768,7 +768,80 @@ func TestBalance_InvalidHash(t *testing.T) {
 	mockGraphQL.AssertExpectations(t)
 }
 
-func TestCall(t *testing.T) {
+func TestCall_GetBlockByNumber(t *testing.T) {
+	mockJSONRPC := &mocks.JSONRPC{}
+	mockGraphQL := &mocks.GraphQL{}
+
+	c := &Client{
+		c:              mockJSONRPC,
+		g:              mockGraphQL,
+		traceSemaphore: semaphore.NewWeighted(100),
+	}
+
+	ctx := context.Background()
+
+	mockJSONRPC.On(
+		"CallContext",
+		ctx,
+		mock.Anything,
+		"eth_getBlockByNumber",
+		"0x2af0",
+		false,
+	).Return(
+		nil,
+	).Run(
+		func(args mock.Arguments) {
+			r := args.Get(1).(*json.RawMessage)
+
+			file, err := ioutil.ReadFile("testdata/block_10992.json")
+			assert.NoError(t, err)
+
+			*r = json.RawMessage(file)
+			assert.NoError(t, (*r).UnmarshalJSON(file))
```

Comment:
@shrimalmadhur I'm having difficulty with this test due to the format. Not sure the best way resolve it. I didn't find a matching type for the response.

**Error:**

```
ok      github.com/coinbase/rosetta-ethereum/services   (cached)
?       github.com/coinbase/rosetta-ethereum/cmd        [no test files]
ok      github.com/coinbase/rosetta-ethereum/configuration      (cached)
--- FAIL: TestCall_GetBlockByNumber (0.00s)
panic: interface conversion: interface {} is *map[string]interface {}, not *json.RawMessage [recovered]
        panic: interface conversion: interface {} is *map[string]interface {}, not *json.RawMessage

goroutine 63 [running]:
testing.tRunner.func1.2({0x10368b500, 0x140001478f0})
        /opt/homebrew/opt/go/libexec/src/testing/testing.go:1209 +0x258
testing.tRunner.func1(0x140004ef1e0)
        /opt/homebrew/opt/go/libexec/src/testing/testing.go:1212 +0x284
panic({0x10368b500, 0x140001478f0})
        /opt/homebrew/opt/go/libexec/src/runtime/panic.go:1038 +0x21c
github.com/coinbase/rosetta-ethereum/ethereum.TestCall_GetBlockByNumber.func1({0x140001ffb00, 0x5, 0x6})
        /Users/ahussein/Development/rosetta/rosetta-ethereum/ethereum/client_test.go:794 +0x224
github.com/stretchr/testify/mock.(*Mock).MethodCalled(0x14000103310, {0x1037fa0c0, 0xb}, {0x140001ffb00, 0x5, 0x6})
        /Users/ahussein/go/pkg/mod/github.com/stretchr/testify@v1.7.0/mock/mock.go:452 +0xc98
github.com/stretchr/testify/mock.(*Mock).Called(0x14000103310, {0x140001ffb00, 0x5, 0x6})
        /Users/ahussein/go/pkg/mod/github.com/stretchr/testify@v1.7.0/mock/mock.go:383 +0x178
github.com/coinbase/rosetta-ethereum/mocks/ethereum.(*JSONRPC).CallContext(0x14000103310, {0x103736ad0, 0x140000a6008}, {0x1036427e0, 0x140000103e8}, {0x1033e11d6, 0x14}, {0x1400039d660, 0x2, 0x2})
        /Users/ahussein/Development/rosetta/rosetta-ethereum/mocks/ethereum/jsonrpc.go:37 +0x21c
github.com/coinbase/rosetta-ethereum/ethereum.(*Client).Call(0x1400035df28, {0x103736ad0, 0x140000a6008}, 0x1400005cf08)
        /Users/ahussein/Development/rosetta/rosetta-ethereum/ethereum/client.go:1207 +0x79c
github.com/coinbase/rosetta-ethereum/ethereum.TestCall_GetBlockByNumber(0x140004ef1e0)
        /Users/ahussein/Development/rosetta/rosetta-ethereum/ethereum/client_test.go:803 +0x3c8
testing.tRunner(0x140004ef1e0, 0x10371eee0)
        /opt/homebrew/opt/go/libexec/src/testing/testing.go:1259 +0x104
created by testing.(*T).Run
        /opt/homebrew/opt/go/libexec/src/testing/testing.go:1306 +0x328
FAIL    github.com/coinbase/rosetta-ethereum/ethereum   0.455s
FAIL
make: *** [test] Error 1
```

Reply:
It's weird that I see some other [tests](https://github.com/coinbase/rosetta-ethereum/blob/master/ethereum/client_test.go#L1071) using the same conversion but not failing. Am I missing something here?

Reply:
I've simplified the test and fixed the lint CI error, but still running into same issue. Wonder if it's to do with the mock function call arguments?

Reply:
So I was just comparing the test case which is passing and failing. The one with passing as `args.Get(1)` as `{interface{}} nil` but one with failing one has `{interface{} | *map[string]interface{}}` - that's the only difference I can see. so probably something to do with mocks only. Investigating more.

Reply:
Since we have no types for this response, I changed the mock Arguments function to 
```
func(args mock.Arguments) {
	r := args.Get(1).(*map[string]interface{})

	file, err := ioutil.ReadFile("testdata/block_10992.json")
	assert.NoError(t, err)

	err = json.Unmarshal(file, r)
	assert.NoError(t, err)
},
```
Not sure if this makes sense. but it does pass the test. I am not that much of an expert in go 😅